02-react-week-1/spreading-and-destructuring.md

```diff
@@ -0,0 +1,379 @@
+
+```javascript
+const student = {
+  name: 'Bob',
+  grades: [3, 4, 5],
+  address: {
+    street: '123 Elm St',
+    city: 'Trenton',
+    state: 'NJ',
+    zip: 11111,
+   }
+};
+
+const newStudent = { ...student };
+
+newStudent.name = 'Alice';
+newStudent.address.street = '4th avenue';
+
+console.log(newStudent);
+console.log('---');
+console.log(student);
+```
+
+The above code outputs: 
+
+```javascript
+{
+  name: 'Alice',
+  grades: [ 12, 4, 5 ],
+  address: {
+    street: '4th avenue',
+    city: 'Trenton',
+    state: 'NJ',
+    zip: 11111
+  }
+}
+---
+{
+  name: 'Bob',
+  grades: [ 12, 4, 5 ],
+  address:
+   { 
+     street: '4th avenue',
+     city: 'Trenton',
+     state: 'NJ',
+     zip: 11111
+  } 
+}
+```
+<br/><br/>
+
+**Question** Look at the code above.  Why do both Alice & Bob have a street address of 4th avenue?
+
+<details>
+  <summary>
+    Click here to see the answer.
+  </summary>
+
+  The `...` operator is making a _shallow_ copy of the original object.  It does not copy all the values of `student`'s properties.  It instead copies the _reference_ to the `address` object and the `grades` array.  To make a _deep copy_ of the original object you would need to do things [differently.](https://medium.com/@tkssharma/objects-in-javascript-object-assign-deep-copy-64106c9aefab)
+</details>
+
+
+## Examining Code
+
+The following code simulates merging new permissions into a user account.  With a partner examine the code and answer the questions:
+
+```javascript
+const combinePreferences = function combinePreferences(user, ...otherPrefs) {
+  otherPrefs.forEach((preference) => {
+    user.preferences = { ...user.preferences, ...preference}
+  });
+  return user;
+}
+
+const user = {
+  name: 'ada',
+  preferences: {
+    superuser: true,
+    billableUser: false,
+  }
+};
+
+const billing = {
+  billableUser: true,
+  getsReports: true,
+  canDeployApps: false,
+};
+
+const project = {
+  canCreateProjects: false,
+  canDeployApps: false,
+  canDeployToTest: true,
+};
+
+console.log(combinePreferences(user, billing, project));
+```
+
+**Questions** 
+
+As a pair read the code and predict the result.  Then run the resulting code in the terminal.
+
+1. Identify where the spread operator and rest parameters are.
+2. Did it output what you expected?  Why?
+3. Is the `forEach` loop necessary?  Could it be done without a loop?  Why or why not?
+
+<details>
+  <summary>
+    Click here to see some answers.
+  </summary>
+
+  1. The rest parameter occurs at `combinePreferences(user, ...preferences)`, and the spread operator is used twice at `user.preferences = { ...user.preferences, ...preference}`.  
+  2. The result is the user object with it's preference attribute updated merging them in order of appearance.
+  3. The `forEach` loop is needed because using the spread operator on an array will result in merging in the array's keys (index numbers) and values into the preferences.  
```

Comment:
```suggestion
  1. The rest parameter occurs at `combinePreferences(user, ...preferences)`, and the spread operator is used twice at `user.preferences = { ...user.preferences, ...preference}`.  
  1. The result is the user object with it's preference attribute updated merging them in order of appearance.
  1. The `forEach` loop is needed because using the spread operator on an array will result in merging in the array's keys (index numbers) and values into the preferences.  
```